src/kernel/pmm.zig

```diff
@@ -0,0 +1,214 @@
+    bitmap = PmmBitmap.init(mem.mem_kb * 1024 / BLOCK_SIZE, allocator) catch @panic("Bitmap allocation failed");
+
+    // Occupy the regions of memory that the memory map describes as reserved
+    for (mem.mem_map) |entry| {
+        if (entry.@"type" != MEMORY_AVAILABLE) {
+            var addr = std.mem.alignBackward(@intCast(usize, entry.addr), BLOCK_SIZE);
+            var end = @intCast(usize, entry.addr + (entry.len - 1));
+            // If the end address can be aligned without overflowing then align it
+            if (end <= std.math.maxInt(usize) - BLOCK_SIZE)
+                end = std.mem.alignForward(end, BLOCK_SIZE);
+            while (addr < end) : (addr += BLOCK_SIZE) {
+                setAddr(addr) catch |e| switch (e) {
+                    // We can ignore out of bounds errors as the memory won't be available anyway
+                    PmmBitmap.BitmapError.OutOfBounds => break,
+                    else => panic(@errorReturnTrace(), "Failed setting address 0x{x} from memory map as occupied: {}", .{ addr, e }),
+                };
+            }
+        }
+    }
+    // Occupy kernel memory
+    var addr = std.mem.alignBackward(@ptrToInt(mem.physaddr_start), BLOCK_SIZE);
+    while (addr < @ptrToInt(mem.physaddr_end)) : (addr += BLOCK_SIZE) {
+        setAddr(addr) catch |e| switch (e) {
+            PmmBitmap.BitmapError.OutOfBounds => panic(@errorReturnTrace(), "Failed setting kernel code address 0x{x} as occupied. The amount of system memory seems to be too low for the kernel image: {}", .{ addr, e }),
+            else => panic(@errorReturnTrace(), "Failed setting kernel code address 0x{x} as occupied: {}", .{ addr, e }),
+        };
+    }
+    log.logInfo("Done\n", .{});
+    if (build_options.rt_test) {
+        runtimeTests(mem);
+    }
+}
+
+fn runtimeTests(mem: *const MemProfile) void {
+    // Make sure that occupied memory can't be allocated
+    var prev_alloc: usize = std.math.maxInt(usize);
+    while (alloc()) |alloced| {
+        if (prev_alloc == alloced) {
+            panic(null, "PMM allocated the same address twice: 0x{x}", .{alloced});
+        }
+        prev_alloc = alloced;
+        for (mem.mem_map) |entry| {
+            if (entry.@"type" != MEMORY_AVAILABLE) {
+                var addr = std.mem.alignBackward(@intCast(usize, entry.addr), BLOCK_SIZE);
+                if (addr == alloced) {
+                    panic(null, "PMM allocated an address that should be reserved by the memory map: 0x{x}", .{addr});
+                }
+            }
+        }
+        if (alloced >= std.mem.alignBackward(@ptrToInt(mem.physaddr_start), BLOCK_SIZE) and alloced < std.mem.alignForward(@ptrToInt(mem.physaddr_end), BLOCK_SIZE)) {
+            panic(null, "PMM allocated an address that should be reserved by kernel code: 0x{x}", .{alloced});
+        }
+    }
+    log.logInfo("PMM: Tested allocation\n", .{});
+}
+
+test "alloc" {
+    bitmap = try Bitmap(u32).init(32, std.heap.direct_allocator);
```

Comment:
Idk whether bitmap needs to be cleared after the test as it is a global. This could be tested at the beginning of each test to ensure the bitmap is clear. As subsequentially tests could fail or produce undefined results.

Reply:
It shouldn't need to be since each test creates its own bitmap so there won't be any re-usage.

Reply:
But `bitmap` is global and if you forget to reinitialise it could be a problem. If you re-assign it for each test then it should be fine, but just to bare it in mind as I had this problem when testing the tty.

Reply:
Indeed. If we look at the existing tests before creating a new one we should spot that the bitmap needs to be recreated.